core/src/link_resolver.rs

```diff
@@ -157,35 +157,36 @@ impl LinkResolverTrait for LinkResolver {
     /// Supports links of the form `/ipfs/ipfs_hash` or just `ipfs_hash`.
     fn cat<'a>(&'a self, logger: &'a Logger, link: &'a Link) -> DynTryFuture<'a, Vec<u8>> {
         async move {
-            // Discard the `/ipfs/` prefix (if present) to get the hash.
-            let path = link.link.trim_start_matches("/ipfs/").to_owned();
-
-            if let Some(data) = self.cache.lock().unwrap().get(&path) {
-                trace!(logger, "IPFS cache hit"; "hash" => &path);
-                return Ok(data.clone());
-            }
-            trace!(logger, "IPFS cache miss"; "hash" => &path);
-
-            let (stat, client) =
-                select_fastest_client_with_stat(&self.clients, &path, self.timeout).await?;
-
-            // FIXME: Having an env variable here is a problem for consensus.
-            // Index Nodes should not disagree on whether the file should be read.
-            let max_file_size: Option<u64> = read_u64_from_env(MAX_IPFS_FILE_SIZE_VAR);
-            restrict_file_size(&path, &stat, &max_file_size)?;
-
-            let path = path.clone();
             let retry_fut = if self.retry {
-                retry("ipfs.cat", &logger).no_limit()
+                retry("object.stat + ipfs.cat", &logger).no_limit()
             } else {
-                retry("ipfs.cat", &logger).limit(1)
+                retry("object.stat + ipfs.cat", &logger).limit(1)
             }
             .timeout(self.timeout);
 
-            let result = retry_fut
+            retry_fut
                 .run(move || {
-                    let path = path.clone();
                     async move {
+                        // Pick the client that a) has the file and b) responds to
+                        // Discard the `/ipfs/` prefix (if present) to get the hash.
+                        let path = link.link.trim_start_matches("/ipfs/").to_owned();
+
+                        if let Some(data) = self.cache.lock().unwrap().get(&path) {
+                            trace!(logger, "IPFS cache hit"; "hash" => &path);
+                            return Ok(data.clone());
+                        }
+                        trace!(logger, "IPFS cache miss"; "hash" => &path);
+
+                        // `object.stat` fastest
+                        let (stat, client) =
+                            select_fastest_client_with_stat(&self.clients, &path, self.timeout)
+                                .await?;
+
+                        // FIXME: Having an env variable here is a problem for consensus.
+                        // Index Nodes should not disagree on whether the file should be read.
+                        let max_file_size: Option<u64> = read_u64_from_env(MAX_IPFS_FILE_SIZE_VAR);
```

Comment:
I'm not sure how much of an edge case this is, but it doesn't make sense to retry on errors that are not intermittent. If restrict_file_size fails, this will retry in vain until the timeout is reached. Our retry function doesn't really give you that option though. (And most of our error handling is the anyhow ad-hoc sort where this would be difficult even if the retry function supported it.

I'd like to take a stab at cleaning up more of our async stuff - especially around diesel and retry. But I don't think we'll have time for it.

Reply:
Yeah, this is bad. I've reverted combining everything into a single retry.